prow/gcsupload/run_test.go

```diff
@@ -0,0 +1,167 @@
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"reflect"
+	"sort"
+	"strings"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/util/diff"
+
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/pod-utils/downwardapi"
+	"k8s.io/test-infra/prow/pod-utils/gcs"
+)
+
+func TestOptions_AssembleTargets(t *testing.T) {
+	var testCases = []struct {
+		name     string
+		options  Options
+		paths    []string
+		extra    map[string]gcs.UploadFunc
+		expected []string
+	}{
+		{
+			name: "no extra paths should upload infra files",
+			options: Options{
+				GCSConfiguration: &kube.GCSConfiguration{
+					PathStrategy: kube.PathStrategyExplicit,
+					Bucket:       "bucket",
+				},
+			},
+			expected: []string{
+				"pr-logs/directory/job/build.txt",
+				"pr-logs/directory/job/latest-build.txt",
+				"pr-logs/pull/org_repo/1/job/latest-build.txt",
+			},
+		},
+		{
+			name: "extra paths should be uploaded under job dir",
+			options: Options{
+				GCSConfiguration: &kube.GCSConfiguration{
+					PathStrategy: kube.PathStrategyExplicit,
+					Bucket:       "bucket",
+				},
+			},
+			extra: map[string]gcs.UploadFunc{
+				"something": gcs.DataUpload(strings.NewReader("data")),
+				"else":      gcs.DataUpload(strings.NewReader("data")),
+			},
+			expected: []string{
+				"pr-logs/pull/org_repo/1/job/build/something",
+				"pr-logs/pull/org_repo/1/job/build/else",
+				"pr-logs/directory/job/build.txt",
+				"pr-logs/directory/job/latest-build.txt",
+				"pr-logs/pull/org_repo/1/job/latest-build.txt",
+			},
+		},
+		{
+			name: "literal files should be uploaded under job dir",
+			options: Options{
+				Items: []string{"something", "else"},
+				GCSConfiguration: &kube.GCSConfiguration{
+					PathStrategy: kube.PathStrategyExplicit,
+					Bucket:       "bucket",
+				},
+			},
+			paths: []string{"something", "else", "notforupload"},
+			expected: []string{
+				"pr-logs/pull/org_repo/1/job/build/something",
+				"pr-logs/pull/org_repo/1/job/build/else",
+				"pr-logs/directory/job/build.txt",
+				"pr-logs/directory/job/latest-build.txt",
+				"pr-logs/pull/org_repo/1/job/latest-build.txt",
+			},
+		},
+		{
+			name: "directories should be uploaded under job dir",
+			options: Options{
+				Items: []string{"something"},
+				GCSConfiguration: &kube.GCSConfiguration{
+					PathStrategy: kube.PathStrategyExplicit,
+					Bucket:       "bucket",
+				},
+			},
+			paths: []string{"something/", "something/else", "notforupload"},
+			expected: []string{
+				"pr-logs/pull/org_repo/1/job/build/something/else",
+				"pr-logs/directory/job/build.txt",
+				"pr-logs/directory/job/latest-build.txt",
+				"pr-logs/pull/org_repo/1/job/latest-build.txt",
+			},
+		},
+	}
+
+	spec := &downwardapi.JobSpec{
+		Job:  "job",
+		Type: kube.PresubmitJob,
+		Refs: kube.Refs{
+			Org:  "org",
+			Repo: "repo",
+			Pulls: []kube.Pull{
+				{
+					Number: 1,
+				},
+			},
+		},
+		BuildId: "build",
+	}
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
```

Comment:
Hmm... you should encourage us to start using this pattern elsewhere

Does this help with testgrid?

Reply:
Go will paralelize better with this but to be honest the biggest reason I chose it was so I could `defer` :)